src/main/scala/net/snowflake/spark/snowflake/SnowflakeRDD.scala

```diff
@@ -81,88 +82,186 @@ private[snowflake] class SnowflakeRDD[T: ClassTag](
-            if (!is256) {
-              stream = getDecryptedStream(stream,
-                                          masterKey,
-                                          dataObject.getObjectMetadata)
+    fileType match {
+      case "csv" => {
+        val reader = new SnowflakeRecordReader
+        val converter = Conversions.createRowConverter[T](resultSchema)
+
+        try {
+          mats.foreach {
+            case (file, queryId, smkId) =>
+              if (queryId != null) {
+                var stream: InputStream = null
+
+                val amazonClient =
+                  createS3Client(is256,
+                    masterKey,
+                    queryId,
+                    smkId,
+                    awsID,
+                    awsKey,
+                    awsToken,
+                    parallel)
+
+                val (bucketName, stagePath) =
+                  extractBucketNameAndPath(stageLocation)
+
+                var stageFilePath = file
+
+                if (!stagePath.isEmpty) {
+                  stageFilePath =
+                    SnowflakeUtil.concatFilePathNames(stagePath, file, "/")
+                }
+
+                val dataObject = amazonClient.getObject(bucketName, stageFilePath)
+                stream = dataObject.getObjectContent
+
+                if (!is256) {
+                  stream = getDecryptedStream(stream,
+                    masterKey,
+                    dataObject.getObjectMetadata)
+                }
+
+                stream = compress match {
+                  case "gzip" => new GZIPInputStream(stream)
+                  case _      => stream
+                }
+
+                reader.addStream(stream)
+              }
+          }
+        } catch {
+          case ex: Exception =>
+            reader.closeAll()
+            SnowflakeConnectorUtils.handleS3Exception(ex)
+        }
+
+        new InterruptibleIterator(context, new Iterator[T] {
+
+          private var finished = false
+          private var havePair = false
+
+          override def hasNext: Boolean = {
+            if (!finished && !havePair) {
+              try {
+                finished = !reader.nextKeyValue
+              } catch {
+                case e: IOException =>
+                  finished = true
+              }
+
+              havePair = !finished
             }
+            !finished
+          }
 
-            stream = compress match {
-              case "gzip" => new GZIPInputStream(stream)
-              case _      => stream
+          override def next(): T = {
+            if (!hasNext) {
+              throw new java.util.NoSuchElementException("End of stream")
             }
+            havePair = false
 
-            reader.addStream(stream)
+            converter(reader.getCurrentValue)
           }
+        })
       }
-    } catch {
-      case ex: Exception =>
-        reader.closeAll()
-        SnowflakeConnectorUtils.handleS3Exception(ex)
-    }
+      case "parquet" => {
```

Comment:
There is a lot of duplicate, reused code here. Can we refactor to make this cleaner?